tests/python/test_slowmo_fsdp.py

```diff
@@ -0,0 +1,382 @@
+        return self.out(F.relu(self.net(x)))
+
+
+class TestCommunicationHooks(FSDPTest):
+    def _init_one_layer_fsdp(self, sharding_strategy):
+        net = torch.nn.Linear(1, 5, bias=False)
+        return FSDP(
+            net,
+            device_id=torch.cuda.current_device(),
+            sharding_strategy=sharding_strategy,
+        ).to(self.rank)
+
+    @skip_if_lt_x_gpu(2)
+    @parametrize("sharding_strategy", [ShardingStrategy.NO_SHARD])
+    def test_slowmo_hook_with_grad_sync(
+        self,
+        sharding_strategy: Optional[ShardingStrategy],
+    ):
+
+        fsdp_net = self._init_one_layer_fsdp(sharding_strategy)
+        inpt = torch.tensor([self.rank]).float()
+        inpt.to(self.rank)  # type: ignore[call-overload]
+
+        slowMoState = slow_momentum_comm.SlowMoState(subgroup=None, grad_sync=True)
+        # check that a default subgroup was created,
+        # for small scale experiments equal to `World_Size`
+        self.assertEqual(slowMoState.subgroup.size(), dist.get_world_size())
+
+        cur_subgroup = dist.new_group(ranks=[self.rank])
+        self.assertEqual(cur_subgroup.size(), 1)
+        slowMoState = slow_momentum_comm.SlowMoState(cur_subgroup, grad_sync=True)
+        # check that state has subgroup registered
+        self.assertEqual(slowMoState.subgroup.size(), cur_subgroup.size())
+        self.assertEqual(slowMoState.subgroup.rank(), 0)
+
+        fsdp_net.register_comm_hook(slowMoState, slow_momentum_comm.slowmo_hook)
+
+        fsdp_net.zero_grad()
+        loss = fsdp_net(inpt).sum()
+        loss.backward()
+
+        # Make sure grads were not reduced,
+        # since each subgroup is only one worker.
+        # Gradient in this case is equal to rank
+        self.assertEqual(fsdp_net.params[0].grad[0], self.rank)
+
+    @skip_if_lt_x_gpu(2)
+    @parametrize("sharding_strategy", [ShardingStrategy.NO_SHARD])
+    def test_slowmo_hook_no_grad_sync(
+        self,
+        sharding_strategy: Optional[ShardingStrategy],
+    ):
+
+        fsdp_net = self._init_one_layer_fsdp(sharding_strategy)
+        inpt = torch.tensor([self.rank]).float()
+        inpt.to(self.rank)  # type: ignore[call-overload]
+
+        # create a subgroup equal to the whole WORLD
+        cur_subgroup = dist.distributed_c10d._get_default_group()
+        self.assertEqual(cur_subgroup.size(), dist.get_world_size())
+        slowMoState = slow_momentum_comm.SlowMoState(cur_subgroup, grad_sync=False)
+        # check that state has subgroup registered
+        self.assertEqual(slowMoState.subgroup.size(), cur_subgroup.size())
+
+        fsdp_net.register_comm_hook(slowMoState, slow_momentum_comm.slowmo_hook)
+        fsdp_net.zero_grad()
+        loss = fsdp_net(inpt).sum()
+        loss.backward()
+
+        # Make sure grads were not reduced, since `grad_sync` is set to False
+        # Gradient in this case is equal to rank
+        self.assertEqual(fsdp_net.params[0].grad[0], self.rank)
```

Comment:
looks like some code duplication from the above test, can we dedupe as much as possible?